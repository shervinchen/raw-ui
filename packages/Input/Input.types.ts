import { InputHTMLAttributes, ChangeEvent, FocusEvent } from 'react';

export type InputTypes = 'default' | 'primary' | 'warning' | 'error';

export type InputSizes = 'sm' | 'md' | 'lg';

export interface BaseInputProps {
  value?: string;
  defaultValue?: string;
  placeholder?: string;
  type?: InputTypes;
  size?: InputSizes;
  width?: string;
  htmlType?: string;
  readOnly?: boolean;
  disabled?: boolean;
  className?: string;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
  onFocus?: (event: FocusEvent<HTMLInputElement>) => void;
  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
  autoComplete?: string;
}

type NativeInputProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  keyof BaseInputProps
>;

export type InputProps = BaseInputProps & NativeInputProps;

export interface InputSizeStyles {
  fontSize?: string;
  height?: string;
}

export interface InputBasicStyles {
  borderColor?: string;
  color?: string;
  backgroundColor?: string;
  cursor?: 'text' | 'not-allowed';
}

export type InputStyles = InputSizeStyles & InputBasicStyles;

export interface InputFocusStyles {
  focusBorderColor?: string;
}
