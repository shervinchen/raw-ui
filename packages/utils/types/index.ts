export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
