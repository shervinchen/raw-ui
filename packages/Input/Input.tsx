import React, {
  forwardRef,
  ChangeEvent,
  FocusEvent,
  useImperativeHandle,
  useRef,
  ComponentPropsWithRef,
} from 'react';
import classNames from 'classnames';
import { InputProps } from './Input.types';
import { useInputCSS } from './Input.styles';
import { useControlled } from '../utils/hooks';
import { InputGroupConfig } from './InputGroup.types';
import { useInputGroupContext } from './input-group-context';

const mergeInputGroupProps = (
  inputProps: InputProps,
  config: InputGroupConfig
): InputProps => {
  if (!config.isInputGroup) return inputProps;
  return {
    ...inputProps,
    size: config.size,
    type: config.type,
    readOnly: config.readOnly,
    disabled: config.disabled,
  };
};

const Input = forwardRef(
  (inputProps: InputProps, ref: ComponentPropsWithRef<'input'>['ref']) => {
    const inputGroupConfig = useInputGroupContext();
    const {
      type = 'default',
      size = 'md',
      width = '100%',
      htmlType = 'text',
      placeholder = '',
      defaultValue = '',
      disabled = false,
      readOnly = false,
      className = '',
      autoComplete = 'off',
      value,
      onChange,
      onBlur,
      onFocus,
      ...restProps
    } = mergeInputGroupProps(inputProps, inputGroupConfig);
    const inputRef = useRef<HTMLInputElement>(null);
    const [internalValue, setInternalValue] = useControlled<string>({
      defaultValue,
      value,
    });
    const { className: resolveClassName, styles } = useInputCSS({
      type,
      size,
      width,
      disabled,
    });
    const classes = classNames(
      'raw-input',
      disabled && 'raw-disabled-input',
      className,
      resolveClassName
    );

    const handleFocus = (event: FocusEvent<HTMLInputElement>) => {
      onFocus?.(event);
    };

    const handleBlur = (event: FocusEvent<HTMLInputElement>) => {
      onBlur?.(event);
    };

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
      setInternalValue(event.target.value);
      onChange?.(event);
    };

    useImperativeHandle(ref, () => inputRef.current as HTMLInputElement);

    return (
      <>
        <input
          ref={inputRef}
          type={htmlType}
          className={classes}
          value={internalValue}
          placeholder={placeholder}
          disabled={disabled}
          readOnly={readOnly}
          onFocus={handleFocus}
          onBlur={handleBlur}
          onChange={handleChange}
          autoComplete={autoComplete}
          {...restProps}
        />
        {styles}
      </>
    );
  }
);

Input.displayName = 'RawInput';

export default Input;
