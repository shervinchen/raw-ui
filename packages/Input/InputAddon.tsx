import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { useInputGroupContext } from './input-group-context';
import { useInputStyles } from './Input.styles';
import { useTheme } from '../Theme/theme-context';
import { RawUITheme } from '../Theme/preset/preset.type';
import { InputAddonProps } from './Input.types';

const InputAddon: FC<PropsWithChildren<InputAddonProps>> = ({
  className,
  children,
  ...resetProps
}) => {
  const theme: RawUITheme = useTheme();
  const { type, size, disabled } = useInputGroupContext();
  const { fontSize } = useInputStyles({ type, size, disabled });
  const classes = classNames('raw-input-addon', className);

  return (
    <div className={classes} {...resetProps}>
      {children}
      <style jsx>{`
        .raw-input-addon {
          width: auto;
          display: flex;
          align-items: center;
          white-space: nowrap;
          pointer-events: none;
          padding: 0 12px;
          font-size: ${fontSize};
          color: ${theme.palette.neutral['500']};
          background-color: ${theme.palette.neutral['50']};
          border: 1px solid ${theme.palette.neutral['200']};
        }
        .raw-input-left-addon {
          border-right: 0;
          border-radius: 6px 0 0 6px;
        }
        .raw-input-right-addon {
          border-left: 0;
          border-radius: 0 6px 6px 0;
        }
      `}</style>
    </div>
  );
};

const InputLeftAddon: FC<InputAddonProps> = ({
  className = '',
  ...resetProps
}) => {
  const classes = classNames('raw-input-left-addon', className);

  return <InputAddon className={classes} {...resetProps} />;
};

const InputRightAddon: FC<InputAddonProps> = ({
  className = '',
  ...resetProps
}) => {
  const classes = classNames('raw-input-right-addon', className);

  return <InputAddon className={classes} {...resetProps} />;
};

export { InputLeftAddon, InputRightAddon };
