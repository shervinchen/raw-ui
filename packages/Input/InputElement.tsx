import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { useInputGroupContext } from './input-group-context';
import { useInputStyles } from './Input.styles';
import { InputElementProps } from './Input.types';

const InputElement: FC<PropsWithChildren<InputElementProps>> = ({
  placement,
  clickable = false,
  className,
  children,
  ...resetProps
}) => {
  const { type, size, disabled } = useInputGroupContext();
  const { fontSize, height, color } = useInputStyles({ type, size, disabled });
  const classes = classNames('raw-input-element', className);

  return (
    <div className={classes} {...resetProps}>
      {children}
      <style jsx>{`
        .raw-input-element {
          position: absolute;
          top: 0;
          ${placement}: 0;
          display: flex;
          align-items: center;
          justify-content: center;
          width: ${height};
          height: ${height};
          font-size: ${fontSize};
          color: ${color};
          pointer-events: ${clickable ? 'auto' : 'none'};
          cursor: ${clickable ? 'pointer' : 'default'};
        }
      `}</style>
    </div>
  );
};

const InputLeftElement: FC<Omit<InputElementProps, 'placement'>> = ({
  className = '',
  ...resetProps
}) => {
  const classes = classNames('raw-input-left-element', className);

  return (
    <InputElement
      data-testid="inputLeftElement"
      className={classes}
      placement="left"
      {...resetProps}
    />
  );
};

const InputRightElement: FC<Omit<InputElementProps, 'placement'>> = ({
  className = '',
  ...resetProps
}) => {
  const classes = classNames('raw-input-right-element', className);

  return (
    <InputElement
      data-testid="inputRightElement"
      className={classes}
      placement="right"
      {...resetProps}
    />
  );
};

export { InputLeftElement, InputRightElement };
