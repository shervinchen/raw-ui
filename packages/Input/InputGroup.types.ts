import { HTMLAttributes } from 'react';
import { InputSizes, InputTypes } from './Input.types';

interface BaseInputGroupProps {
  className?: string;
  size?: InputSizes;
  type?: InputTypes;
  readOnly?: boolean;
  disabled?: boolean;
}

type NativeInputGroupProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseInputGroupProps
>;

export type InputGroupProps = BaseInputGroupProps & NativeInputGroupProps;

export interface InputGroupConfig {
  size?: InputSizes;
  type?: InputTypes;
  readOnly?: boolean;
  disabled?: boolean;
  isInputGroup: boolean;
}
