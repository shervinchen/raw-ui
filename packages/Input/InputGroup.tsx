import React, { FC, PropsWithChildren, cloneElement, useMemo } from 'react';
import classNames from 'classnames';
import { InputGroupConfig, InputGroupProps } from './InputGroup.types';
import { InputGroupContext } from './input-group-context';
import { getValidChildren } from '../utils/common';
import { useInputStyles } from './Input.styles';
import Input from './Input';
import { InputLeftElement, InputRightElement } from './InputElement';
import { InputLeftAddon, InputRightAddon } from './InputAddon';

const useComputedInputStyles = ({ type, size, disabled }: InputGroupProps) => {
  const { height, horizontalPadding } = useInputStyles({
    type,
    size,
    disabled,
  });
  const style = {
    paddingLeft: horizontalPadding,
    paddingRight: horizontalPadding,
    borderTopRightRadius: '6px',
    borderBottomRightRadius: '6px',
    borderTopLeftRadius: '6px',
    borderBottomLeftRadius: '6px',
  };
  const styles = [
    {
      type: InputLeftElement,
      newStyle: {
        paddingLeft: height,
      },
    },
    {
      type: InputRightElement,
      newStyle: {
        paddingRight: height,
      },
    },
    {
      type: InputLeftAddon,
      newStyle: {
        borderTopLeftRadius: '0px',
        borderBottomLeftRadius: '0px',
      },
    },
    {
      type: InputRightAddon,
      newStyle: {
        borderTopRightRadius: '0px',
        borderBottomRightRadius: '0px',
      },
    },
  ];
  return {
    style,
    styles,
  };
};

const InputGroup: FC<PropsWithChildren<InputGroupProps>> = ({
  className = '',
  size = 'md',
  type = 'default',
  readOnly = false,
  disabled = false,
  children,
  ...resetProps
}) => {
  const initialConfig = useMemo<InputGroupConfig>(
    () => ({
      size,
      type,
      readOnly,
      disabled,
      isInputGroup: true,
    }),
    [disabled, readOnly, size, type]
  );

  const classes = classNames('raw-input-group', className);

  const useComputedInputStyle = () => {
    const { style, styles } = useComputedInputStyles({
      type,
      size,
      disabled,
    });
    let computedStyle = style;
    getValidChildren(children).forEach((child) => {
      const result = styles.find((item) => item.type === child.type);
      if (result) {
        const { newStyle } = result;
        computedStyle = { ...computedStyle, ...newStyle };
      }
    });
    return computedStyle;
  };

  const computedInputStyle = useComputedInputStyle();

  const cloneChildren = getValidChildren(children).map((child) => {
    return child.type !== Input
      ? child
      : cloneElement(child, { style: computedInputStyle });
  });

  return (
    <InputGroupContext.Provider value={initialConfig}>
      <div className={classes} {...resetProps}>
        {cloneChildren}
        <style jsx>{`
          .raw-input-group {
            width: 100%;
            display: inline-flex;
            position: relative;
          }
        `}</style>
      </div>
    </InputGroupContext.Provider>
  );
};

export default InputGroup;
