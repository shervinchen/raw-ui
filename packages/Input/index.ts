import Input from './Input';
import InputGroup from './InputGroup';
import { InputLeftElement, InputRightElement } from './InputElement';
import { InputLeftAddon, InputRightAddon } from './InputAddon';

export type InputComponentType = typeof Input & {
  Group: typeof InputGroup;
  LeftElement: typeof InputLeftElement;
  RightElement: typeof InputRightElement;
  LeftAddon: typeof InputLeftAddon;
  RightAddon: typeof InputRightAddon;
};

(Input as InputComponentType).Group = InputGroup;
(Input as InputComponentType).LeftElement = InputLeftElement;
(Input as InputComponentType).RightElement = InputRightElement;
(Input as InputComponentType).LeftAddon = InputLeftAddon;
(Input as InputComponentType).RightAddon = InputRightAddon;

export default Input as InputComponentType;
