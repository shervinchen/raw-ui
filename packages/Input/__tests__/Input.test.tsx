import React, { ChangeEvent, useState } from 'react';
import { fireEvent, render, renderHook } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from '..';
import { InputProps, InputSizes, InputTypes } from '../Input.types';
import { useInputFocusStyles, useInputStyles } from '../Input.styles';

const typeColorMap = {
  primary: '#000000',
  success: '#0070f3',
  warning: '#f5a623',
  error: '#ee0000',
};

const sizeHeightMap = {
  sm: '32px',
  md: '40px',
  lg: '48px',
};

describe('Input', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Input />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should forward ref', () => {
    const ref = React.createRef<HTMLInputElement>();
    const { container } = render(<Input ref={ref} />);
    const input = container.firstChild;
    expect(input).toEqual(ref.current);
  });

  test('should support custom class name', () => {
    const { container } = render(<Input className="custom-input" />);
    expect(container.firstChild).toHaveClass('custom-input');
  });

  test('should support type text', async () => {
    const { container } = render(<Input />);
    const input = container.firstChild as Element;
    await userEvent.type(input, 'test text');
    expect(input).toHaveValue('test text');
  });

  test('should support placeholder', () => {
    const { container } = render(<Input placeholder="Placeholder..." />);
    const input = container.firstChild as Element;
    expect(input).toHaveAttribute('placeholder', 'Placeholder...');
  });

  test('should support blur event', () => {
    const onBlur = jest.fn();
    const { container } = render(<Input onBlur={onBlur} />);
    const input = container.firstChild as Element;
    fireEvent.blur(input);
    expect(onBlur).toHaveBeenCalledTimes(1);
  });

  test('should support focus event', () => {
    const onFocus = jest.fn();
    const { container } = render(<Input onFocus={onFocus} />);
    const input = container.firstChild as Element;
    fireEvent.focus(input);
    expect(onFocus).toHaveBeenCalledTimes(1);
  });

  test('should support uncontrolled value', () => {
    const { container } = render(<Input defaultValue="default value" />);
    const input = container.firstChild as Element;
    expect(input).toHaveValue('default value');
  });

  test('should support controlled value', () => {
    const onChange = jest.fn();

    const Component = (props: InputProps) => {
      const [value, setValue] = useState('');

      return (
        <Input
          value={value}
          onChange={(event: ChangeEvent<HTMLInputElement>) => {
            setValue(event.target.value);
            props.onChange?.(event);
          }}
        />
      );
    };
    const { container } = render(<Component onChange={onChange} />);
    const input = container.firstChild as Element;
    expect(input).toHaveValue('');
    fireEvent.change(input, { target: { value: 'test text' } });
    expect(input).toHaveValue('test text');
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  ['primary', 'success', 'warning', 'error'].forEach(
    (item: Exclude<InputTypes, 'default'>) => {
      test(`should render ${item} type`, () => {
        const { container } = render(<Input type={item} />);
        const input = container.firstChild as Element;
        expect(getComputedStyle(input).borderColor).toBe(typeColorMap[item]);
      });
    }
  );

  ['sm', 'md', 'lg'].forEach((item: InputSizes) => {
    test(`should render ${item} size`, () => {
      const { container } = render(<Input size={item} />);
      const input = container.firstChild as Element;
      expect(getComputedStyle(input).height).toBe(sizeHeightMap[item]);
    });
  });

  test('should support custom width', () => {
    const { container } = render(<Input width="160px" />);
    const input = container.firstChild as Element;
    expect(getComputedStyle(input).width).toBe('160px');
  });

  test('should support disabled', () => {
    const onChange = jest.fn();
    const { container } = render(<Input disabled onChange={onChange} />);
    const input = container.firstChild as Element;
    expect(input).toBeDisabled();
    expect(input).toHaveValue('');
    fireEvent.change(input, { target: { value: 'test text' } });
    expect(input).toHaveValue('');
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  test('should support readonly', () => {
    const onChange = jest.fn();
    const { container } = render(<Input readOnly onChange={onChange} />);
    const input = container.firstChild as Element;
    expect(input).toHaveAttribute('readonly');
    expect(input).toHaveValue('');
    fireEvent.change(input, { target: { value: 'test text' } });
    expect(input).toHaveValue('');
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  test('should support inside element', () => {
    const { getByText } = render(
      <Input.Group>
        <Input.LeftElement>$</Input.LeftElement>
        <Input placeholder="Enter amount" />
        <Input.RightElement>.0</Input.RightElement>
      </Input.Group>
    );

    expect(getByText('$')).toBeInTheDocument();
    expect(getByText('.0')).toBeInTheDocument();
  });

  test('should inside element support clickable style', () => {
    const { container } = render(
      <Input.Group>
        <Input.LeftElement clickable>$</Input.LeftElement>
        <Input placeholder="Enter amount" />
        <Input.RightElement>.0</Input.RightElement>
      </Input.Group>
    );
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    const leftElement = container.querySelector(
      '.raw-input-left-element'
    ) as Element;
    expect(getComputedStyle(leftElement).pointerEvents).toBe('auto');
    expect(getComputedStyle(leftElement).cursor).toBe('pointer');
  });

  test('should support addon element', () => {
    const { getByText } = render(
      <Input.Group>
        <Input.LeftAddon>https://</Input.LeftAddon>
        <Input placeholder="your domain" />
        <Input.RightAddon>.com</Input.RightAddon>
      </Input.Group>
    );

    expect(getByText('https://')).toBeInTheDocument();
    expect(getByText('.com')).toBeInTheDocument();
  });

  test('should addon element support custom class name', () => {
    const { container } = render(
      <Input.Group>
        <Input.LeftAddon className="left-addon">https://</Input.LeftAddon>
        <Input placeholder="your domain" />
        <Input.RightAddon className="right-addon">.com</Input.RightAddon>
      </Input.Group>
    );

    expect(container.querySelector('.left-addon')).toBeInTheDocument();
    expect(container.querySelector('.right-addon')).toBeInTheDocument();
  });

  test('should get default style when type is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useInputStyles({
        type: 'unknown' as InputTypes,
        size: 'md',
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useInputStyles({
        type: undefined as unknown as InputTypes,
        size: 'md',
        disabled: false,
      })
    );
    expect(result1.current.borderColor).toBe('#eaeaea');
    expect(result2.current.borderColor).toBe('#eaeaea');
  });

  test('should get md size when size is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useInputStyles({
        type: 'default',
        size: 'unknown' as InputSizes,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useInputStyles({
        type: 'default',
        size: undefined as unknown as InputSizes,
        disabled: false,
      })
    );
    expect(result1.current.height).toBe('40px');
    expect(result2.current.height).toBe('40px');
  });

  test('should get default focus style when type is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useInputFocusStyles({
        type: 'unknown' as InputTypes,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useInputFocusStyles({
        type: undefined as unknown as InputTypes,
        disabled: false,
      })
    );
    expect(result1.current.focusBorderColor).toBe('#666666');
    expect(result2.current.focusBorderColor).toBe('#666666');
  });
});
