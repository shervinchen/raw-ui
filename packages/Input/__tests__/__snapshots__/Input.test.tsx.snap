// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match the snapshot 1`] = `
<DocumentFragment>
  <input
    autocomplete="off"
    class="raw-input jsx-2920133267"
    placeholder=""
    type="text"
    value=""
  />
</DocumentFragment>
`;
